src/comm/AcceptLimiter.cc

```diff
@@ -22,38 +22,20 @@ Comm::AcceptLimiter::Instance()
 }
 
 void
-Comm::AcceptLimiter::defer(const Comm::TcpAcceptor::Pointer &afd)
+Comm::AcceptLimiter::defer(const AsyncCall::Pointer &call)
 {
-    debugs(5, 5, afd->conn << "; already queued: " << deferred_.size());
-    deferred_.push_back(afd);
-}
-
-void
-Comm::AcceptLimiter::removeDead(const Comm::TcpAcceptor::Pointer &afd)
-{
-    for (auto it = deferred_.begin(); it != deferred_.end(); ++it) {
-        if (*it == afd) {
-            *it = nullptr; // fast. kick() will skip empty entries later.
-            debugs(5,4, "Abandoned client TCP SYN by closing socket: " << afd->conn);
-            return;
-        }
-    }
-    debugs(5,4, "Not found " << afd->conn << " in queue, size: " << deferred_.size());
+    debugs(5, 5, call << "; already queued: " << deferred_.size());
+    deferred_.push_back(call);
 }
 
 void
 Comm::AcceptLimiter::kick()
 {
     debugs(5, 5, "size=" << deferred_.size());
-    while (deferred_.size() > 0 && Comm::TcpAcceptor::okToAccept()) {
-        /* NP: shift() is equivalent to pop_front(). Giving us a FIFO queue. */
-        TcpAcceptor::Pointer temp = deferred_.front();
+    if (deferred_.size() > 0 && Comm::TcpAcceptor::okToAccept()) {
+        auto call = deferred_.front();
         deferred_.erase(deferred_.begin());
-        if (temp.valid()) {
-            debugs(5, 5, "doing one.");
-            temp->acceptNext();
-            break;
-        }
+        ScheduleCallHere(call);
     }
```

Comment:
The problem I am having with that for now is lack of AsyncCall API info to tell if the Call is actually dead. The ScheduleCalHere() always returns true, and as you point out the FDs count does not decrease until later.

It should be possible to check how many FDs are free on each kick(). Such that repeated stall conditions try and schedule ever more Calls from the queue per kick(). But thats just a workaround and may not help if the load stays high.

Reply:
FWIW, this same lack of ability to identify dead Calls is why removeDead() is dropped. It simply cannot be implemented.

Reply:
> The problem I am having with that for now is lack of AsyncCall API info to tell if the Call is actually dead. The ScheduleCalHere() always returns true, and as you point out the FDs count does not decrease until later.

Adding AsyncCall::actuallyDead() API does not help because the TcpAcceptor job may disappear after the not-yet-dead call is successfully scheduled. The job may even disappear after the scheduled call is successfully fired (but before it accepts the connection, consuming the free FD "slot"). In the "worst" case, the job may disappear after the job accepts the connection _and_ closes the descriptor, effectively leaving the FD "slot" free as if the job never existed.

See my earlier [comment](https://github.com/squid-cache/squid/pull/857#issuecomment-875976489) for a solution that works despite all those complications.

> It should be possible to check how many FDs are free on each kick().

Yes, computing that value is easy, but knowing that value does not help solving the problem. And the solution I sketched would not benefit from knowing that value, so I have not suggested computing it (IIRC).

> FWIW, this same lack of ability to identify dead Calls is why removeDead() is dropped. It simply cannot be implemented.

A removeDead() call can be implemented -- it is possible to remove a given AsyncCall object from a set of deferred calls, of course. Squid already has similar code in comm_remove_close_handler(AsyncCall::Pointer), but that function is a bit more complex because it is more general than removeDead().

Please note that removeDead(x) is very different from removeAllDead(void). We can and should keep the former, adjusting it from job pointers to call pointers. I am not suggesting to add the latter.

Reply:
Resolved already using your alternative design.